### ISSUE TITLE
ODEnet produces NaN when evaluating

### ISSUE THREAD

**Issue by DrChainsaw** [DrChainsaw/neuralODE4j#5]
This happens when evaluating after having completed a training epoch. Does not seem to happen if number of training examples is reduced which makes it pretty annoying to debug. Smallest number of examples when behaviour is reproduced was 80 iterations of 128 examples.

Furthermore, if sysouts are added to print variables which are suspects of being the source of NaNs, the problem seems to go away, at least with the above number of examples. This holds for more or less any variable, so it did not work as a method to narrow down the error further.

Due to this, current plan is to debug this with low effort during off hours and focus on #1.

Tested: 
* Remove all special workspace handling in OdeVertex and just use the provided LayerWorkspaceMgr
* Various permutations of scope out of workspaces
* Detaching activations, inputs and epsilons.
* Running with CPU backend for a few number of iterations to test if a workspace error is caught there which is not caught with the CUDA backend (there was actually an issue with batchnorm putting state in workspace for ArrayType.INPUT, but fixing it didn't solve this issue)
* Run without asynch prefetch
* Implement #1 and hope problem just goes away
* Running whole model with workspace mode NONE for training and inference (don't know why I didn't think of this first)
* Serializing the model before eval and see if same issue persists with a deserialized graph. Probably requires #3 
* Running validation with training = true by just always giving "true" in constructor. This prevents the issue from surfacing and accuracy seems sensible. Drawback is that batchnorm is most likely not as effective as it should be. See post below about root cause.

Not tested:
* Try another data set
* ...

**Comment by DrChainsaw** [comment]
Cause of NaN when evaluating is that batchnorm has negative variance. Reason why this does not show up in training is that the parameter for variance ("var") is not used in training, only the batch variance is used.

Will try to train without batchnorm layers to see if parameters for other layers are somewhat ok (they should be as they are used during training).

**Comment by DrChainsaw** [comment]
Batchnorm gradient for variance seems to be incorrect when computed by the ODE solver, taking negative steps well beyond the rolling average of the variances.  

As a workaround, it is possible to set training = true always when creating ForwardPass in OdeVertex#doForward.

**Comment by DrChainsaw** [comment]
Prime suspect root cause: Dl4j updates the sliding mean and variance of batchnorm by giving the delta from the EWMA filter as a gradient. This does not seem to play nicely with the ODE solver as it fiddles with the gradient. 

The solver seems to consistently flip the sign of the mean and var "gradients" and a quick fix to just flip the sign of the mean and var gradients makes the NaNs go away for the first few epochs (haven't tested for longer).

A real fix would probably be to mask mean and var gradients from the solver somehow.
